### PR TITLE
Fix report widget

### LINKED ISSUE #1369: [bug] playerReportBug
When I try to 'Report Coordinate' in game, I am immediately disconnected.

When checking the file 'data/reports/PlayerName.txt' the log has been created, but without the message typed by the player.

------------------------------
Name: PlayerName [Position X: 1000 Y: 1000 Z: 7]
Bug Report:

The 'Bug Report' field is empty even though I've typed in game.

I tried to check the C++ code but I could not solve, just realized that this line is disconnecting me:
```
	if (msg.isOverrun ()) {
		disconnect ();
	}
```
File: protocolgame.cpp Line 549 to 551

Does anyone have any idea how to solve this issue?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/16499691-bug-playerreportbug?utm_campaign=plugin&utm_content=tracker%2F392151&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F392151&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by WibbenZ**
Closes #1369

### CODE REVIEW

**Comment by paulosalvatore** [comment]
Brilliant!

**Comment by kornholi** [comment]
> It is now posible for players to report bugs aswell

There's a reason it is disabled... it would be too easy for someone to fill up your hard drive with junk reports

**Comment by WibbenZ** [comment]
@kornholi reverted it, thing is that it won't be used otherwise.
And it would require alot of reports to fill a hard drive up, but I do get it if we are talking about servers like shadowcores etc with alot of players.

**Comment by paulosalvatore** [comment]
Is it possible to concatenate category value into the file?
Something like:
```
fprintf(file, "------------------------------\nName: %s [Position: %u, %u, %u]\nBug Report: %s\nCategory: %s", player->getName().c_str(), position.x, position.y, position.z, message.c_str(), category.c_str());
```

**Comment by ranisalt** [comment]
I suggest you use a more modern approach with iostream/fstream and  string streams to write to a file instead.

**Comment by WibbenZ** [comment]
Well I don't really see how that would be that much better, but if @marksamman or @kornholi  agrees with you then sure.

**Comment by marksamman** [comment]
Don't read player->getPosition() in parseBugReport, and include both map position and player position in the report if there's a map position.

I disagree about fstream, please revert the fstream changes and squash your commits.

**Comment by WibbenZ** [comment]
It seems to return the player position, when I tested it they all returned the same position or what do you mean by map position?
Only the map report returns a position (target tile position, so you can report bugs from a distance)

**Comment by marksamman** [comment]
When there's a map report, the map position and the player position should be included in the report. For other types of reports, only the players position should be included. The players position should **not** be read in parseBugReport.

**Comment by WibbenZ** [comment]
I can agree partly on that, but in some cases like npcs that have diffrent scripts and diffrent locations but the same name (Morris? in Yalahar), would make it harder unless they give a full report message of where they are and what the message was.

IMO we should log the map position in all cases but if it's a map but we also log the player position.

**Comment by marksamman** [comment]
I have no idea what you're talking about or what relevance NPC's and towns have in this.

What I'm asking you to do is simply this:
If it's a map report, include both the map position and the players position in the report.
If it's not a map report, include the players position in the report.

**Comment by WibbenZ** [comment]
Have no clue why it dosen't squash them, am I doing something wrong?

````
git rebase -i head~7
```

I then get this in my notepad
````
pick 386b6dc Fix report widget
pick 200445e Revert posibility for players to report
pick 121603d Fstream
pick 71a2575 Ups :p
pick 166bb9c Fix z position message, aswell as indentation
pick 7f517c9 Github did not like my indentation ;o
pick 5082e89 Revert fstream, position changes

# Rebase bd12849..5082e89 onto bd12849
#
# Commands:
#  p, pick = use commit
#  r, reword = use commit, but edit the commit message
#  e, edit = use commit, but stop for amending
#  s, squash = use commit, but meld into previous commit
#  f, fixup = like "squash", but discard this commit's log message
#  x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out


```

**Comment by ninjalulz** [comment]
Replace all pick (except the first one "Fix report widget") with squash.

**Comment by ranisalt** [comment]
You may also use `git rebase -i master`, you do not need to count commits.

Also, `s` is enough instead of `squash`, e.g.
```
pick 386b6dc Fix report widget
s 200445e Revert posibility for players to report
s 121603d Fstream
s 71a2575 Ups :p
s 166bb9c Fix z position message, aswell as indentation
s 7f517c9 Github did not like my indentation ;o
s 5082e89 Revert fstream, position changes
```

**Comment by WibbenZ** [comment]
Failed above I guess? xD

```
pick 121603d Fstream
squash 71a2575 Ups :p
squash 166bb9c Fix z position message, aswell as indentation
squash 7f517c9 Github did not like my indentation ;o
squash 5082e89 Revert fstream, position changes
squash 0438718 squash
squash 7e5288f Revert "squash"

# Rebase 200445e..7e5288f onto 200445e
#
# Commands:
#  p, pick = use commit
#  r, reword = use commit, but edit the commit message
#  e, edit = use commit, but stop for amending
#  s, squash = use commit, but meld into previous commit
#  f, fixup = like "squash", but discard this commit's log message
#  x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out


```

And I get this, but github does not update?
```
Successfully rebased and updated refs/heads/master
```

**Comment by ranisalt** [comment]
Yes, that's because of git's distributed nature. You have now to force a push with `git push -f` to send updates to GitHub.

**Comment by WibbenZ** [comment]
When I used git push -f it just said "Everything is up-to-date", complicated stuff haha.
Why not just merge it as-is?

**Comment by ranisalt** [comment]
Because you have lots of commits fixing your own comits with useless and uninformative messages, that's just polluting the history.

[Check this tutorial](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) and try again, else redo the changes on a new branch. You can ease the burden creating a patch, then applying on master.

**Comment by WibbenZ** [comment]
Ive tested it about 10 times and ive checked the tutorial and it still does not work.
It does not seem to find the file unless I use git add (with the file the squash info is in) but it still does not update, if I run it again it says "Everything is up-to-date".

What should I do? :p

**Comment by gugahoa** [comment]
Open a new pull request, if you need.

**Comment by WibbenZ** [comment]
Would be nice to know how to do this in the future, not having to open more then one pull xD

**Comment by ninjalulz** [comment]
@WibbenZ 

Try an alternate squash method
```
git reset --soft HEAD~9
git add -A
git commit -m "Fix report widget"
git push -f
```

**Comment by WibbenZ** [comment]
@ninjalulz seems to work "kinda" haha :p 
In the "git windows" program I can now only see one commit, but it has other stuff that have been fixed aswell (Maybe it should have that?)
And ive tried to sync it but it dosen't update here on the website.
All it says is "Already up-to-date" "Everything is up-to-date".

**Comment by WibbenZ** [comment]
Ehm.. it seems to have moved to my other commit, noticed it now: https://github.com/otland/forgottenserver/pull/1381

**Comment by ninjalulz** [comment]
You should have switched to your master branch before executing those commands. xD

```git checkout master```

**Comment by WibbenZ** [comment]
Hahahha there we go :D 
Thanks @ninjalulz  <3, ill write this up, been trying to google it aswell but nothing worked haha...

**Comment by peonso** [comment]
> There's a reason it is disabled... it would be too easy for someone to fill up your hard drive with junk reports

@kornholi a cooldown could be set for players.

@WibbenZ as you presented in otland, and I agree, it's better to have the report saved in the database, easier to manage by aac or even ingame. Having it disabled for players really hurt the functionality of it.

**Comment by WibbenZ** [comment]
Ye, but I guess that is how they wanted it and that is the reason they haven't changed it.
On my phone atm so it's annoying to check but the function below this one had a comment saying it should be moved to sql insted of file logging it.
So they probbly have a reason for it.

**Comment by WibbenZ** [comment]
Anything left to do here? @marksamman

File: src/game.cpp [hunk 1]
```diff
@@ -4855,8 +4855,13 @@ void Game::playerReportBug(uint32_t playerId, const std::string& bug)
 	std::string fileName = "data/reports/" + player->getName() + " report.txt";
 	FILE* file = fopen(fileName.c_str(), "a");
 	if (file) {
-		const Position& position = player->getPosition();
-		fprintf(file, "------------------------------\nName: %s [Position X: %u Y: %u Z: %u]\nBug Report: %s\n", player->getName().c_str(), position.x, position.y, position.z, bug.c_str());
+		const Position playerPosition = player->getPosition();
```
Comment (marksamman) [inline src/game.cpp]: Make this a reference.

**Comment by WibbenZ** [comment]
Done

File: src/game.cpp [hunk 2]
```diff
@@ -4855,8 +4855,13 @@ void Game::playerReportBug(uint32_t playerId, const std::string& bug)
 	std::string fileName = "data/reports/" + player->getName() + " report.txt";
 	FILE* file = fopen(fileName.c_str(), "a");
 	if (file) {
-		const Position& position = player->getPosition();
-		fprintf(file, "------------------------------\nName: %s [Position X: %u Y: %u Z: %u]\nBug Report: %s\n", player->getName().c_str(), position.x, position.y, position.z, bug.c_str());
+		const Position& playerPosition = player->getPosition();
+		if (category == BUG_CATEGORY_MAP) {
+			fprintf(file, "------------------------------\nName: %s [Map Position: %u, %u, %u] [Player Position: %u, %u, %u]\nnComment: %s\n", player->getName().c_str(), position.x, position.y, position.z, playerPosition.x, playerPosition.y, playerPosition.z, message.c_str());
```
Comment (brunominervino) [inline src/game.cpp]: \nnComment:

**Comment by marksamman** [comment]
Thanks!

**Pull request merged**